tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -190,6 +236,361 @@ void BuildDotClusters(Block* block,
+      auto lhs_type = dot.lhs().getType().dyn_cast<RankedTensorType>();
+      auto concat_dim_size = lhs_type.getDimSize(concat_dim);
+      if (concat_dim_size == -1) {
+        is_dynamic_shape = true;
+      } else {
+        concat_dim_sum += concat_dim_size;
+      }
+      lhs_operands.push_back(dot.lhs());
+    }
+    //concat lhs
+    auto lhs_rank = orig_lhs_type.getRank();
+    SmallVector<int64_t, 4> lhs_shapes(lhs_rank, ShapedType::kDynamicSize);
+    for (int64_t i = 0; i < lhs_rank; i++) {
+      if (i == concat_dim) {
+        if (is_dynamic_shape) {
+          lhs_shapes[i] = -1;
+        } else {
+          lhs_shapes[i] = concat_dim_sum;
+        }
+      } else {
+        lhs_shapes[i] = orig_lhs_type.getDimSize(i);  
+      } 
+    }
+    auto lhs_type =
+        RankedTensorType::get(lhs_shapes, orig_lhs_type.getElementType());
+    lhs = builder.create<mhlo::ConcatenateOp>(loc, lhs_type, lhs_operands,
+                                                    concat_dim);
+    rhs = foremost_dot.rhs();
+  } else {
+    assert("share_type error");
+  }
+  
+  // Concat Result type.
+  auto result_rank = orig_result_type.getRank();
+  SmallVector<int64_t, 4> result_shapes(result_rank, ShapedType::kDynamicSize);
+  for (int64_t i = 0; i < result_rank; i++) {
+    if (i == concat_dim) {
+      if (is_dynamic_shape) {
+        result_shapes[i] = -1;
+      } else {
+        result_shapes[i] = concat_dim_sum;
+      }
+    } else {
+      result_shapes[i] = orig_result_type.getDimSize(i);  
+    }   
+  }
+  auto result_type =
+      RankedTensorType::get(result_shapes, orig_result_type.getElementType());
+  // Build dot dimension numbers.
+  auto dim_numbers = foremost_dot.dot_dimension_numbers();
+
+  SmallVector<int64_t> lhs_batching_dims;
+  auto lhs_batch = dim_numbers.getLhsBatchingDimensions();
+  lhs_batching_dims.insert(lhs_batching_dims.end(), lhs_batch.begin(),
+                           lhs_batch.end());
+
+  SmallVector<int64_t> rhs_batching_dims;
+  auto rhs_batch = dim_numbers.getRhsBatchingDimensions();
+  rhs_batching_dims.insert(rhs_batching_dims.end(), rhs_batch.begin(),
+                           rhs_batch.end());
+
+  SmallVector<int64_t> lhs_contracting_dims;
+  auto lhs_contract = dim_numbers.getLhsContractingDimensions();
+  for (auto& val : lhs_contract) {
+    lhs_contracting_dims.push_back(val);
+  }
+
+  SmallVector<int64_t> rhs_contracting_dims;
+  auto rhs_contract = dim_numbers.getRhsContractingDimensions();
+  for (auto& val : rhs_contract) {
+    rhs_contracting_dims.push_back(val);
+  }
+
+  auto dot_dimension_attr = mhlo::DotDimensionNumbersAttr::get(
+      builder.getContext(), lhs_batching_dims, rhs_batching_dims,
+      lhs_contracting_dims, rhs_contracting_dims);
+  
+  // Build new DotGeneralOp
+  Value merged_dot = builder.create<mhlo::DotGeneralOp>(
+      loc, result_type, lhs, rhs, dot_dimension_attr, nullptr);
+  // Build slice  for each of the original dot op, and replace the dot.  
```

Comment:
Nit: 1) an extra space in the comment; 2) add line between function blocks and remove line between the comment and its corresponding function block.